### ISSUE TITLE
Chaining a single group will not wait for group member completion

### ISSUE THREAD

**Issue by Enteee** [celery/celery#8182]
# Checklist

- [x] I have verified that the issue exists against the `main` branch of Celery.
- [x] This has already been asked to the [discussions forum](https://github.com/celery/celery/discussions) first. -> https://github.com/celery/celery/discussions/8170
- [x] I have read the relevant section in the
  [contribution guide](https://docs.celeryq.dev/en/main/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/main)
  to find out if the bug was already fixed in the main branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `main` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**:

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
software -> celery:5.3.0b2 (dawn-chorus) kombu:5.3.0b3 py:3.10.7
            billiard:4.1.0 py-amqp:5.1.1
platform -> system:Linux arch:64bit, ELF
            kernel version:5.15.94-1.qubes.fc32.x86_64 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:pyamqp results:redis:///

broker_url: 'amqp://guest:********@localhost:5672//'
result_backend: 'redis:///'
deprecated_settings: None
```

</p>
</details>

# Steps to Reproduce

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
amqp==5.1.1
asttokens==2.2.1
async-timeout==4.0.2
backcall==0.2.0
billiard==4.1.0
celery @ git+https://github.com/celery/celery.git@2e168af493fb741e9ddc7261780a0070927351ed
click==8.1.3
click-didyoumean==0.3.0
click-plugins==1.1.1
click-repl==0.2.0
decorator==5.1.1
executing==1.2.0
flower==1.2.0
humanize==4.6.0
ipdb==0.13.13
ipython==8.11.0
jedi==0.18.2
kombu==5.3.0b3
matplotlib-inline==0.1.6
parso==0.8.3
pexpect==4.8.0
pickleshare==0.7.5
prometheus-client==0.16.0
prompt-toolkit==3.0.38
ptyprocess==0.7.0
pure-eval==0.2.2
Pygments==2.14.0
pytz==2022.7.1
redis==4.5.1
six==1.16.0
stack-data==0.6.2
tomli==2.0.1
tornado==6.2
traitlets==5.9.0
vine==5.0.0
watchdog==2.3.1
wcwidth==0.2.6
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
#!/usr/bin/env python3
from random import randint
from time import sleep
from celery import Celery, Signature, chain, chord, group
from celery.utils.log import  get_task_logger

app2 = Celery('app2', backend='redis://', broker='pyamqp://')
logger = get_task_logger(__name__)

@app2.task
def t1(*args):
    logger.info(f"{args[-1]}({args[:-1]})")
    return args

@app2.task
def t2(*args):
    logger.info(f"start {args[-1]}({args[:-1]})")
    sleep(randint(1, 5))
    logger.info(f"end {args[-1]}({args[:-1]})")
    return args

if __name__ == "__main__":
    chain(
        group(
            t2.s("group1"),
            chain(
                group(
                    t1.s("group2->chain1->group1"),
                    t1.s("group2->chain1->group2"),
                ),
            ),
        ),
        t1.s("chord"),
    ).delay()
```

</p>
</details>

# Expected Behavior

<details><summary>The chord runs **after** the group1 task finishes:</summary>
<p>

```
[2023-03-31 10:59:39,365: INFO/ForkPoolWorker-2] Task app2.t2[592939f3-22bb-4e76-9b28-9a70b90a4f05] succeeded in 5.018022193999059s: ('group1',)
[2023-03-31 10:59:39,365: INFO/MainProcess] Task app2.t1[ad551c2a-e7a0-4191-9eea-983b83f99ada] received
[2023-03-31 10:59:39,366: INFO/ForkPoolWorker-2] app2.t1[ad551c2a-e7a0-4191-9eea-983b83f99ada]: chord(([['group1'], [['group2->chain1->group1'], ['group2->chain1->group2']]],))
[2023-03-31 10:59:39,367: INFO/ForkPoolWorker-2] Task app2.t1[ad551c2a-e7a0-4191-9eea-983b83f99ada] succeeded in 0.0007412640006805304s: ([['group1'], [[...], [...]]], 'chord')
```

</p>
</details> 

# Actual Behavior

The chord runs **before** the group1 task finishes:

```
[2023-03-31 10:40:25,661: INFO/MainProcess] Task app2.t1[4e447663-d92e-46fe-923a-3199b353df28] received
[2023-03-31 10:40:25,662: INFO/ForkPoolWorker-3] app2.t1[4e447663-d92e-46fe-923a-3199b353df28]: chord(([None, [['group2->chain1->group1'], ['group2->chain1->group2']]],))
[2023-03-31 10:40:25,663: INFO/ForkPoolWorker-3] Task app2.t1[4e447663-d92e-46fe-923a-3199b353df28] succeeded in 0.0013457320019369945s: ([None, [[...], [...]]], 'chord')
[2023-03-31 10:40:27,098: INFO/ForkPoolWorker-2] app2.t2[89265d0d-a876-4ce8-9e1d-d7f68a38246d]: end group1(())
[2023-03-31 10:40:27,101: INFO/ForkPoolWorker-2] Task app2.t2[89265d0d-a876-4ce8-9e1d-d7f68a38246d] succeeded in 2.0062880830009817s: ('group1',)
```

**Comment by Enteee** [comment]
I have implemented a small workaround for this problem:

```python3
#!/usr/bin/env python3
from random import randint
from time import sleep
from celery import Celery, Signature, chain, chord, group
from celery.utils.log import  get_task_logger

app2 = Celery('app2', backend='redis://', broker='pyamqp://')
logger = get_task_logger(__name__)

@app2.task
def t1(*args):
    logger.info(f"{args[-1]}({args[:-1]})")
    return args

@app2.task
def t2(*args):
    logger.info(f"start {args[-1]}({args[:-1]})")
    sleep(randint(1, 5))
    logger.info(f"end {args[-1]}({args[:-1]})")
    return args

@app2.task
def completer(results, *args):
    """Task that does nothing, can be used in a chord to complete the chord"""
    return results

def chainable_group(*tasks, **options) -> Signature:
    """Make a group that can be chained to another task"""
    return chord(group(*tasks, **options), completer.s())

if __name__ == "__main__":
    chain(
        group(
            t2.s("group1"),
            chain(
                chainable_group(
                    t1.s("group2->chain1->group1"),
                    t1.s("group2->chain1->group2"),
                ),
            ),
        ),
        t1.s("chord"),
    ).delay()

```

<details><summary>Which produces the following output</summary>
<p>

```
[2023-03-31 10:59:32,167: INFO/MainProcess] celery@dev ready.
[2023-03-31 10:59:34,344: INFO/MainProcess] Task app2.t2[592939f3-22bb-4e76-9b28-9a70b90a4f05] received
[2023-03-31 10:59:34,347: INFO/ForkPoolWorker-2] app2.t2[592939f3-22bb-4e76-9b28-9a70b90a4f05]: start group1(())
[2023-03-31 10:59:34,349: INFO/MainProcess] Task app2.t1[4eda9150-0f6e-4b8b-9f1f-f68238c9b818] received
[2023-03-31 10:59:34,350: INFO/MainProcess] Task app2.t1[6c02b51d-4215-4601-a97b-22496152ed15] received
[2023-03-31 10:59:34,351: INFO/ForkPoolWorker-1] app2.t1[6c02b51d-4215-4601-a97b-22496152ed15]: group2->chain1->group2(())
[2023-03-31 10:59:34,351: INFO/ForkPoolWorker-3] app2.t1[4eda9150-0f6e-4b8b-9f1f-f68238c9b818]: group2->chain1->group1(())
[2023-03-31 10:59:34,353: INFO/ForkPoolWorker-1] Task app2.t1[6c02b51d-4215-4601-a97b-22496152ed15] succeeded in 0.00289611900007003s: ('group2->chain1->group2',)
[2023-03-31 10:59:34,355: WARNING/ForkPoolWorker-3] No hostname was supplied. Reverting to default 'localhost'
[2023-03-31 10:59:34,361: INFO/MainProcess] Task app2.completer[f1a85198-d56b-4668-b296-f8a2fa20cdbe] received
[2023-03-31 10:59:34,361: INFO/ForkPoolWorker-3] Task app2.t1[4eda9150-0f6e-4b8b-9f1f-f68238c9b818] succeeded in 0.011034281000320334s: ('group2->chain1->group1',)
[2023-03-31 10:59:34,363: INFO/ForkPoolWorker-3] Task app2.completer[f1a85198-d56b-4668-b296-f8a2fa20cdbe] succeeded in 0.001131998000346357s: [['group2->chain1->group1'], ['group2->chain1->group2']]
[2023-03-31 10:59:35,871: INFO/MainProcess] Events of group {task} enabled by remote.
[2023-03-31 10:59:39,352: INFO/ForkPoolWorker-2] app2.t2[592939f3-22bb-4e76-9b28-9a70b90a4f05]: end group1(())
[2023-03-31 10:59:39,356: WARNING/ForkPoolWorker-2] No hostname was supplied. Reverting to default 'localhost'
[2023-03-31 10:59:39,365: INFO/ForkPoolWorker-2] Task app2.t2[592939f3-22bb-4e76-9b28-9a70b90a4f05] succeeded in 5.018022193999059s: ('group1',)
[2023-03-31 10:59:39,365: INFO/MainProcess] Task app2.t1[ad551c2a-e7a0-4191-9eea-983b83f99ada] received
[2023-03-31 10:59:39,366: INFO/ForkPoolWorker-2] app2.t1[ad551c2a-e7a0-4191-9eea-983b83f99ada]: chord(([['group1'], [['group2->chain1->group1'], ['group2->chain1->group2']]],))
[2023-03-31 10:59:39,367: INFO/ForkPoolWorker-2] Task app2.t1[ad551c2a-e7a0-4191-9eea-983b83f99ada] succeeded in 0.0007412640006805304s: ([['group1'], [[...], [...]]], 'chord')
```


</p>
</details>

**Comment by Enteee** [comment]
I do struggle to reproduce this with the newest version of celery.

**Comment by auvipy** [comment]
this should have been fixed in v5.6.0

**Comment by Enteee** [comment]
Sorry, I was too early. I can still reproduce this with `5.6.2 (recovery)`

Here's another reproducible example:

```python
#!/usr/bin/env python3.11
import multiprocessing
from time import sleep

import celery
from celery import Celery, chord, group, chain
from kombu import serialization
from celery.utils.log import get_task_logger


logger = get_task_logger(__name__)

app = Celery("tasks", broker="amqp://XXXXXXX", backend="redis://XXXXXXXXX")
app.conf.accept_content = ["json", "pickle"]
app.conf.result_accept_content = ["json", "pickle"]
app.conf.task_serializer = "pickle"
app.conf.result_serializer = "pickle"
serialization.register_pickle()
serialization.enable_insecure_serializers()

# Additional settings from production
app.conf.task_acks_late = True
app.conf.task_track_started = True
app.conf.task_default_queue = "celery_bug"
app.conf.task_default_queue_type = "quorum"
app.conf.worker_detect_quorum_queues = True


# === Test tasks ===


@app.task
def short_running(*args):
    logger.info(f"{args}")
    return args


@app.task
def long_running(*args):
    logger.info(f"start {args}")
    sleep(5)
    logger.info(f"end {args}")
    return args


@app.task()
def test_task():
    chain(
        group(
            long_running.s("group1"),
            chain(
                group(
                    short_running.s("group2->chain1->group1"),
                    short_running.s("group2->chain1->group2"),
                ),
            ),
        ),
        short_running.s("chord_completer"),
    ).delay()


def worker_process():
    app.Worker(loglevel="info", concurrency=4).start()


if __name__ == "__main__":
    worker = multiprocessing.Process(target=worker_process)
    worker.start()
    sleep(3)

    try:
        # group() is now patched to chord(group(), __completer.s())
        sig = test_task.s()
        result = sig.delay()
        result.get(timeout=10)
    finally:
        worker.terminate()
        worker.join()

```

When i run this for the first time (`celery_bug` Queue is empty). I do get:

```
 
 -------------- celery@... v5.6.2 (recovery)
--- ***** ----- 
-- ******* ---- Linux-6.12.63-x86_64-with-glibc2.40 2026-03-13 08:59:32
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         tasks:0x7ff2c6769750
- ** ---------- .> transport:   amqp://guest:**@rabbit.loom:5672//
- ** ---------- .> results:     redis://redis.loom/
- *** --- * --- .> concurrency: 4 (prefork)
-- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
--- ***** ----- 
 -------------- [queues]
                .> celery_bug       exchange=celery_bug(direct) key=celery_bug
                

[tasks]
  . tasks.long_running
  . tasks.short_running
  . tasks.test_task

[2026-03-13 08:59:33,159: INFO/Process-1] Connected to amqp://guest:**@rabbit.loom:5672//
[2026-03-13 08:59:33,164: INFO/Process-1] mingle: searching for neighbors
[2026-03-13 08:59:34,195: INFO/Process-1] mingle: sync with 2 nodes
[2026-03-13 08:59:34,195: INFO/Process-1] mingle: sync complete
[2026-03-13 08:59:34,203: INFO/Process-1] Global QoS is disabled. Prefetch count in now static.
[2026-03-13 08:59:34,309: WARNING/Process-1] Exchange celery_bug is a direct exchange and native delayed delivery do not support direct exchanges.
ETA tasks published to this exchange will block the worker until the ETA arrives.
[2026-03-13 08:59:34,325: INFO/Process-1] celery@.... ready.
[2026-03-13 08:59:35,773: ERROR/Process-1] Refusing to deserialize untrusted content of type pickle (application/x-python-serialize)
[2026-03-13 08:59:35,805: ERROR/Process-1] Refusing to deserialize untrusted content of type pickle (application/x-python-serialize)
[2026-03-13 08:59:36,052: INFO/Process-1] Task tasks.test_task[7385d87d-2183-484b-9da3-86d992fd4602] received
[2026-03-13 08:59:36,077: INFO/Process-1] Task tasks.long_running[f905cf71-99e4-4afa-a5fb-7bf499d03b4c] received
[2026-03-13 08:59:36,084: INFO/ForkPoolWorker-3] tasks.long_running[f905cf71-99e4-4afa-a5fb-7bf499d03b4c]: start ('group1',)
[2026-03-13 08:59:36,085: INFO/ForkPoolWorker-1] Task tasks.test_task[7385d87d-2183-484b-9da3-86d992fd4602] succeeded in 0.032373079913668334s: None

worker: Warm shutdown (MainProcess)
[2026-03-13 08:59:36,087: INFO/Process-1] Task tasks.short_running[5dd73679-800c-4e90-8672-9e7b7f774264] received
[2026-03-13 08:59:36,088: INFO/Process-1] Task tasks.short_running[e7a1b36b-748e-4308-bfbf-0b56aee052ce] received
[2026-03-13 08:59:41,084: INFO/ForkPoolWorker-3] tasks.long_running[f905cf71-99e4-4afa-a5fb-7bf499d03b4c]: end ('group1',)
[2026-03-13 08:59:41,090: INFO/ForkPoolWorker-3] Task tasks.long_running[f905cf71-99e4-4afa-a5fb-7bf499d03b4c] succeeded in 5.011941776960157s: ('group1',)
```

-> The `short_running.s("chord_completer")` Never actually runs now. It's even worse with `v5.6.2` as the task is not run too early, but just never runs at all. 

I now also get the following errors:

```
[2026-03-13 08:59:35,773: ERROR/Process-1] Refusing to deserialize untrusted content of type pickle (application/x-python-serialize)
[2026-03-13 08:59:35,805: ERROR/Process-1] Refusing to deserialize untrusted content of type pickle (application/x-python-serialize)
```

**Comment by Enteee** [comment]
Sorry, I was too early. I can still reproduce this with 5.6.2 (recovery)

Here's another minimal reproducible example:

```python
#!/usr/bin/env python3.11
import multiprocessing
from time import sleep

from celery import Celery, chord, group, chain
from kombu import serialization
from celery.utils.log import get_task_logger


logger = get_task_logger(__name__)

app = Celery("tasks", broker="amqp://XXXXXXX", backend="redis://XXXXXX")
app.conf.accept_content = ["json", "pickle"]
app.conf.result_accept_content = ["json", "pickle"]
app.conf.task_serializer = "pickle"
app.conf.result_serializer = "pickle"
serialization.register_pickle()
serialization.enable_insecure_serializers()

# Additional settings from production
app.conf.task_acks_late = True
app.conf.task_track_started = True
app.conf.task_default_queue = "celery_bug"
app.conf.task_default_queue_type = "quorum"
app.conf.worker_detect_quorum_queues = True


# === Test tasks ===


@app.task()
def short_running(*args):
    logger.info(f"{args}")
    return args


@app.task()
def long_running(*args):
    logger.info(f"start {args}")
    sleep(5)
    logger.info(f"end {args}")
    return args


def worker_process():
    app.Worker(loglevel="info", concurrency=4).start()


if __name__ == "__main__":
    worker = multiprocessing.Process(target=worker_process)
    worker.start()
    sleep(3)

    try:
        sig = chain(
            group(
                long_running.s("group1"),
                chain(
                    group(
                        short_running.s("group2->chain1->group1"),
                        short_running.s("group2->chain1->group2"),
                    ),
                ),
            ),
            short_running.s("chord_completer"),
        )
        result = sig.delay()
        result.get(timeout=10)
    finally:
        worker.terminate()
        worker.join()
```

output:

```
 
 -------------- celery@XXXXXXXX v5.6.2 (recovery)
--- ***** ----- 
-- ******* ---- Linux-6.12.63-x86_64-with-glibc2.40 2026-03-13 09:46:35
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         tasks:0x7f70a3b8a0d0
- ** ---------- .> transport:   amqp://guest:**@XXXXXXX:5672//
- ** ---------- .> results:     redis://XXXXXXX/
- *** --- * --- .> concurrency: 4 (prefork)
-- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
--- ***** ----- 
 -------------- [queues]
                .> celery_bug       exchange=celery_bug(direct) key=celery_bug
                

[tasks]
  . tasks.long_running
  . tasks.short_running

[2026-03-13 09:46:35,339: INFO/Process-1] Connected to amqp://guest:**@XXXXX:5672//
[2026-03-13 09:46:35,345: INFO/Process-1] mingle: searching for neighbors
[2026-03-13 09:46:36,368: INFO/Process-1] mingle: sync with 1 nodes
[2026-03-13 09:46:36,369: INFO/Process-1] mingle: sync complete
[2026-03-13 09:46:36,371: INFO/Process-1] Global QoS is disabled. Prefetch count in now static.
[2026-03-13 09:46:36,530: WARNING/Process-1] Exchange celery_bug is a direct exchange and native delayed delivery do not support direct exchanges.
ETA tasks published to this exchange will block the worker until the ETA arrives.
[2026-03-13 09:46:36,547: INFO/Process-1] celery@puddle ready.
[2026-03-13 09:46:37,257: ERROR/Process-1] Refusing to deserialize untrusted content of type pickle (application/x-python-serialize)
[2026-03-13 09:46:38,117: INFO/Process-1] Task tasks.long_running[37fa966d-7aa8-4a80-9e7c-e636116720bc] received
[2026-03-13 09:46:38,123: INFO/ForkPoolWorker-1] tasks.long_running[37fa966d-7aa8-4a80-9e7c-e636116720bc]: start ('group1',)
[2026-03-13 09:46:38,123: INFO/Process-1] Task tasks.short_running[03200309-4239-4c1e-b7fa-a0609d1953ee] received
[2026-03-13 09:46:38,127: INFO/Process-1] Task tasks.short_running[f5064cbe-7b95-4c0d-ac84-c3b72f4f6401] received
[2026-03-13 09:46:38,129: INFO/ForkPoolWorker-3] tasks.short_running[03200309-4239-4c1e-b7fa-a0609d1953ee]: ('group2->chain1->group1',)
[2026-03-13 09:46:38,131: INFO/ForkPoolWorker-3] Task tasks.short_running[03200309-4239-4c1e-b7fa-a0609d1953ee] succeeded in 0.006928347982466221s: ('group2->chain1->group1',)
[2026-03-13 09:46:38,133: INFO/ForkPoolWorker-2] tasks.short_running[f5064cbe-7b95-4c0d-ac84-c3b72f4f6401]: ('group2->chain1->group2',)
[2026-03-13 09:46:38,647: INFO/ForkPoolWorker-2] Task tasks.short_running[f5064cbe-7b95-4c0d-ac84-c3b72f4f6401] succeeded in 0.5191854969598353s: ('group2->chain1->group2',)
[2026-03-13 09:46:38,652: INFO/Process-1] Task tasks.short_running[cd18345e-7621-40bd-bf4a-ae133a0b757e] received
[2026-03-13 09:46:38,654: INFO/ForkPoolWorker-3] tasks.short_running[cd18345e-7621-40bd-bf4a-ae133a0b757e]: ([None, [('group2->chain1->group1',), ('group2->chain1->group2',)]], 'chord_completer')
[2026-03-13 09:46:38,655: INFO/ForkPoolWorker-3] Task tasks.short_running[cd18345e-7621-40bd-bf4a-ae133a0b757e] succeeded in 0.0018690930446609855s: ([None, [(...,), (...,)]], 'chord_completer')

worker: Warm shutdown (MainProcess)
[2026-03-13 09:46:39,263: ERROR/Process-1] Refusing to deserialize untrusted content of type pickle (application/x-python-serialize)
[2026-03-13 09:46:43,123: INFO/ForkPoolWorker-1] tasks.long_running[37fa966d-7aa8-4a80-9e7c-e636116720bc]: end ('group1',)
[2026-03-13 09:46:43,134: INFO/ForkPoolWorker-1] Task tasks.long_running[37fa966d-7aa8-4a80-9e7c-e636116720bc] succeeded in 5.015893026953563s: ('group1',)

```

* I now get `Refusing to deserialize untrusted content of type pickle (application/x-python-serialize)` Errors. This is new in `5.6.2`
* `Task tasks.short_running[cd18345e-7621-40bd-bf4a-ae133a0b757e] succeeded in 0.0018690930446609855s: ([None, [(...,), (...,)]], 'chord_completer')` appears **BEFORE** `[2026-03-13 09:46:43,134: INFO/ForkPoolWorker-1] Task tasks.long_running[37fa966d-7aa8-4a80-9e7c-e636116720bc] succeeded in 5.015893026953563s: ('group1',)` -> Task ordering issue.
* The `result.get(timeout=10)` returns too early. Causing the worker to shutdown while the `tasks.long_running` is still running.

**Comment by Enteee** [comment]
@auvipy : can you please remove the milestone, as this was not fixed in 5.6.0

**Comment by simonepozzoli-pix4d** [comment]
Could you try using chords instead of group and chain?
```python
sig = chord(
    [
        long_running.s("group1"),
        group(
            short_running.s("group2->chain1->group1"),
            short_running.s("group2->chain1->group2"),
        ),
    ],
    short_running.s("chord_completer"),
)
```
Also, I don't understand what `chain(group([A, B]))` should do, why do you need a chain with a single element?

**Comment by Enteee** [comment]
> Also, I don't understand what chain(group([A, B])) should do, why do you need a chain with a single element?

@simonepozzoli-pix4d : I tried to build a minimal reproducible example, stripping away all the fluff from my pipeline which is not required to trigger the issue. 

In my opinion, `chain(group([A, B]))` should behave exactly the same as `group([A, B])`

I also tried:

> ```python
> sig = chord(
>    [
>        long_running.s("group1"),
>        group(
>            short_running.s("group2->chain1->group1"),
>            short_running.s("group2->chain1->group2"),
>        ),
>    ],
>    short_running.s("chord_completer"),
>)
>```

This works as expected.

Following up on your input @simonepozzoli-pix4d , this does work:

```python
        sig = chain(
            group(
                long_running.s("group1"),
                chain(
                    group(
                        short_running.s("group2->chain1->group1"),
                        short_running.s("group2->chain1->group2"),
                    ),
                    short_running.s("group2->chain1"),  # chain after
                ),
            ),
            short_running.s("chord_completer"),
        )
```

but this does not work:

```python
        sig = chain(
            group(
                long_running.s("group1"),
                chain(
                    short_running.s("group2->chain1"),  # chain before
                    group(
                        short_running.s("group2->chain1->group1"),
                        short_running.s("group2->chain1->group2"),
                    ),
                ),
            ),
            short_running.s("chord_completer"),
        )
```

The latter is more close to what i do in my actual pipeline.

**Comment by Enteee** [comment]
~Hang on, i just noticed that I accidentally tested against 5.5.3 not 5.6.2. I am re-testing the my latest comment at the moment.~
The issue described in https://github.com/celery/celery/issues/8182#issuecomment-4054842547 also shows in 5.6.2

**Comment by c32-4e3-844-41ed9dc5d** [comment]
I'm currently investigating the issue and found that the following does also work:
```python
chain(
    group(
        long_running.s("group1"),
        chain(
            short_running.s("group2->chain1"),  # chain before
            group(
                short_running.s("group2->chain1->group1"),
                short_running.s("group2->chain1->group2"),
            ),
            short_running.s("dummy task"),  # workaround
        ),
    ),
    short_running.s("chord_completer"),
)
```
I've also created these two tests which demonstrate two interesting behaviors which are both unexpected / related to this issue:

```python
@app.task()
def short_running(*args):
    return args if args[-1] is True else args[-1]


@app.task()
def long_running(*args):
    sleep(2)
    return args if args[-1] is True else args[-1]

# TODO this never completes??? Chain with single group is strange but this behavior is even stranger
def test_chain_nested_group_single_chain_element():
    sig = chain(
        group(
            long_running.s(1),
            chain(
                group(
                    short_running.s(-2),
                    short_running.s(-3),
                ),
                short_running.s(-4),
                group(
                    short_running.s(5),
                    long_running.s(6),
                    short_running.s(7),
                ),
            ),
            short_running.s(8),
        ),
    )

    # E           celery.exceptions.TimeoutError: The operation timed out.
    result = sig.delay().get(timeout=GET_TIMEOUT)
    expected = [1, [5, 6, 7], 8]
    assert result == expected


# TODO here the last task gets None for the long_running.s(1) instead of the expected 1
# I'm still trying to pin down what the root cause is, any inputs from people who know the celery internals are welcome
def test_chain_nested_group_single_chain_element_foo():
    sig = chain(
        group(
            long_running.s(1),
            chain(
                short_running.s(-2),
                group(
                    short_running.s(3),
                    short_running.s(4),
                ),
            ),
        ),
        short_running.s(5, True),
    )

    result = sig.delay().get(timeout=GET_TIMEOUT)
    # E       AssertionError: assert ([None, [3, 4]], 5, True) == ([1, [3, 4]], 5, True)
    # E
    # E         At index 0 diff: [None, [3, 4]] != [1, [3, 4]]
    expected = ([1, [3, 4]], 5, True)
    assert result == expected
```

**Comment by c32-4e3-844-41ed9dc5d** [comment]
> ```
>     * I now get `Refusing to deserialize untrusted content of type pickle (application/x-python-serialize)` Errors. This is new in `5.6.2`
> 
>     * `Task tasks.short_running[cd18345e-7621-40bd-bf4a-ae133a0b757e] succeeded in 0.0018690930446609855s: ([None, [(...,), (...,)]], 'chord_completer')` appears **BEFORE** `[2026-03-13 09:46:43,134: INFO/ForkPoolWorker-1] Task tasks.long_running[37fa966d-7aa8-4a80-9e7c-e636116720bc] succeeded in 5.015893026953563s: ('group1',)` -> Task ordering issue.
> 
>     * The `result.get(timeout=10)` returns too early. Causing the worker to shutdown while the `tasks.long_running` is still running.

@Enteee the first point is resolved by setting `app.conf.event_serializer = "pickle"`, but I can reproduce your other two points see my comment above, which (in my opinion) shows the same behavior in a unit test which is easier to test than reading logfiles.

**Comment by simonepozzoli-pix4d** [comment]
Most probably the issue is related to Celery not correctly setting the chain or chord options to the tasks inside the final group in the chain, it's possible that the group_id is overridden somehow when the chain is uplifted to a chord (this is just a guess, I haven't tested anything).
Could you try adding a final task in each chain that just forwards the results of previous tasks? To verify that the issue is in fact a group used as last step in a chain that is part of an outer chain.
For example:
```python
@app.task
def ident(*args):
    return args[0]
```
and set it as last step in all the chains.

**Comment by c32-4e3-844-41ed9dc5d** [comment]
> Most probably the issue is related to Celery not correctly setting the chain or chord options to the tasks inside the final group in the chain, it's possible that the group_id is overridden somehow when the chain is uplifted to a chord (this is just a guess, I haven't tested anything). Could you try adding a final task in each chain that just forwards the results of previous tasks? To verify that the issue is in fact a group used as last step in a chain that is part of an outer chain. For example:
> 
> @app.task
> def ident(*args):
>     return args[0]
> 
> and set it as last step in all the chains.

Hey @simonepozzoli-pix4d , I think that could be the problem. I already have [a comment about your workaround](https://github.com/c32-4e3-844-41ed9dc5d/celery/blob/e9620188813c01753419d96155d2a103ed64be4d/t/integration/test_canvas.py#L241-L242) in my merge request tests, adding a task at the end of the inner chain indeed fixes the issue:
```py
c = chain(
    group(
        foo.s(),
        chain(
            foo.s(),
            group(
                foo.s(),
                foo.s(),
            ),
            ident.s(),  # <-- fixes the problem
        ),
    ),
    identity.s()
)
``` 

I'm happy for any inputs on how to tackle this, the canvas internals aren't that simple to get started :) Some AI "thought" / probably hallucinated it has to do with some callbacks and completion counters being wrongfully passed down because a chain will delegate it's callbacks to the last element, and if that's a group said group will delegate the callback to all it's children - but I've no way to rate that theory.

**Comment by simonepozzoli-pix4d** [comment]
@c32-4e3-844-41ed9dc5d I think the culprit is the group implementation that does not work well when combined with other canvas primitives. A workaround could be replacing the current implementation of `group` with a chord that uses `ident.s()` as body:
```python
class group(chord):
    def __init__(header):
        super().__init__(header, ident.s())
```
So that all canvas options set to a group would be set the ident body of the chord.
Alternatively, the chord and chain options should be set in a separate dictionary key, not at root level of task options. So that a task that is part of group can write its result to multiple groups with different group_id even when the task is also part of a chord header. For example:
```python
task.options = {
  "chord": {  # should contain all chord related options
    "group_id": ...,
    "group_index": 0,
    "body": <the signature to execute as chord body>,
  },
  "group_id": ...,  # if the task is also part of a group
}
```
But this might be tricky, it might not be easy to set the correct chord group_index to those tasks since the whole group result should have a single index. The workaround seems easier to implement but it might be a breaking change for some people because groups will effectively become chords.

**Comment by c32-4e3-844-41ed9dc5d** [comment]
Thanks for your input @simonepozzoli-pix4d , just using your heavily simplified group implementation is lacking the `tasks`, so it's not an easy drop-in replacement, but I'll try to get something like that working. Maybe alternatively, we have to always "upgrade" groups to chords instead of only when we have a prev_task, I tried that in my MR and it fixed some timeout errors but also added some new ones, I'll continue with my analysis next week

**Comment by ac2-d2c-c47-4ad4608af** [comment]
Hey, @simonepozzoli-pix4d @auvipy 

I have been tracing through the flow of the above non-working example

```python
sig = chain(
                 group(
                       long_running.s("L"),
                       chain(
                            short_running.s("S_0"),
                            group(
                                short_running.s("S_1"),
                                short_running.s("S_2"),
                            ),
                      ),
                 ),
                short_running.s("E"),
          )
```
(used a slight relabelling), and made some observations which might be useful to solve this issue.
 
In detail, in the outer chain the group and last task are internally converted to a chord, with header signature
```python
group((tasks.long_running('L'), short_running('S_0') | group((short_running('S_1'), short_running('S_2')))))
```
and body
```python
tasks.final_running('E')
```
The execution starts from the chord and the chord header group freezing results in a
```python
<class 'celery.canvas.group'>
(group_id, [<AsyncResult: uuid1>, <GroupResult: group_id [uuid2, uuid3]>])
```
 object. Note that the outer and inner group share the same `group_id`. 

The execution goes on into the execution of the outer group. There, in the method
`_apply_task` of the group class the chord size is stored in the backend via 
```python
app.backend.set_chord_size(group_id, chord_size)
```
At this point of the execution of the example, the `chord_size` it is set to 3, which is correct. It
is important to note that the `chord_size` is computed here freshly from zero, via
the `_chord._descend` method. Note also that the `chord_size` set here for a `group_id`
is the condition in the backend for releasing the corresponding chord barrier.

The execution goes on and submits the inner chain as a ChainMap.

So far so good.

However, what happens now in the `ForkPoolWorker` process is that for one of
the group executions, in the group `_apply_task` the `chord_size` is set again in the backend,
but this time to 2, as presumably the outer tasks are not seen anymore. The 2
overwrites the previously stored `chord_size` of 3 in the backend. Therefore, as soon as the two short inner tasks finish, the chord barrier is released, yielding the observed incorrect ordering.

In contrast, for the working example
```python
sig = chain(
                 group(
                       long_running.s("L"),
                       chain(
                            short_running.s("S_0"),
                            group(
                                short_running.s("S_1"),
                                short_running.s("S_2"),
                            ),
                           short_running.s("S_3"),
                      ),
                 ),
                short_running.s("E"),
          )
```
this does not occur, because in the inner chain the group and last task will be
converted to another chord. The inner chord will be executed in the `ForkPoolWorker` and acquires a new `group_id`. 

I do not know the original motivation of computing and setting the `chord_size` in the group method `_apply_task`. Naively, I would compute and set it instead in the call to the backend `apply_chord`, which will be called only once for each chord as
far as I see it, and therefore the barrier counting would stay consistent.
